Rewrite New-TemporaryFile in C# #2732

kittholland opened the pull request:
Literal translation of New-TemporaryFile to C# cmdlet.

msftclas commented:
Hi __@kittholland__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

powercode commented:
And the command to the UNIX version of the psd1 to fix the build error.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,47 @@
+                    WriteError(errorRecord);
+                    return;
+                }
+            FileInfo file = new FileInfo(filePath);
```
Please correct indentation.
`filePath` can be `null` here and raise exception.
Original cmdlet wrap this:
```C#
if (filePath)
{
   FileInfo file = new FileInfo(filePath);
   WriteObject(file); 
}
```

kittholland replied:
Thanks, I wasn't sure what condition could fall into that, that would not trip the catch return. Thanks for clarifying.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,47 @@
+
+/********************************************************************++
+Copyright (c) Microsoft Corporation.  All rights reserved.
+--********************************************************************/
```
@PowerShellTeam Is it correct? What is a policy about copyright?

kittholland replied:
First cmdlet I've writte and I saw that at the top of every other file, wasn't really sure what the protocol was. It's probably wrong. Removing now.

iSazonov replied:
@kittholland :-) My question was not addressed to you. Please don't rush to remove.

kittholland commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+    }
+}
```
Done.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,46 @@
+                            e,
+                            "NewTemporaryFileWriteError",
+                            ErrorCategory.WriteError,
+                            System.Environment.GetEnvironmentVariable("TEMP")));
```
Repeating code. Please use a variable.

kittholland replied:
Removed duplicate GetEnvironmentVariable call. Thanks

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,46 @@
+                            System.Environment.GetEnvironmentVariable("TEMP")));
+                    return;
+                }
+                //filePath may be null, only writeobject if there is a value
```
We usually use a space after `//`.
And the comment is obvious and likely unnecessary.

iSazonov commented:
@kittholland Thanks for your contribution!

My question about the copyright header was not addressed to you. Please don't rush to remove it.
With this in mind, the PR is LGTM. Please wait review from PG.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// returns a TemporaryFile
+        /// </summary>
+        protected override void BeginProcessing()
```
Putting the logic of a simple cmdlet in the `end` block does protect against stupid pipelines, e.g.

```powershell
function f { begin { write-host 'in begin f'}}
function g { end { write-host 'in end g'}}
g | f
```

This will print
```
in begin f
in end g
```
which is a little weird that the logic for a downstream cmdlet finishes before the upstream command (note no objects are being piped).

In others, you could more easily write some obfuscated (as in, hard to figure out the real logic) pipelines if you use `begin` instead of `end`.

daxian-dbw replied:
Thanks @lzybkr, this is a great example. So it looks the guideline should be using `EndProcessing` (or `End` block in script) for simple logic that doesn't require pipeline input.

daxian-dbw replied:
I think the reason we have those cmdlets with main logic in `Begin`/`BeginProcessing` is that the pattern was copied from other existing cmdlets, without knowing why it should be put in `Begin`/`BeginProcessing`.

@kittholland could you please move the main logic to `EndProcessing` method given the discussion here? Thanks!

kittholland replied:
Yes I will move it.

I left as is because this is an existing cmdlet and did not want to potentially alter any existing implementations (and I did not feel that I could fully comprehend what those possible scenarios would have been).

daxian-dbw replied:
Yep, totally understood. Thanks!

iSazonov replied:
This implies that we need to open a new Issue with analysing what cmdlets must be corrected.

JamesWTruher commented:
Are the tests in `test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1` sufficient, or should there be more?

iSazonov commented:
It would be nice to test i/o on the file.

kittholland commented:
I am happy to add an additional test regarding IO but I am uncertain if testing IO after the new file was generated is within the scope of this command.

I could validate additional attributes of the resulting file, for example that the extension is tmp, that the length is 0, and that it was created during the span of testing, but I'm not certain there are any conditions that would ever cause those tests to fail.

iSazonov commented:
I believe that there is no need to examine the underlying level. We need to check only the cmdlet.
```powershell
$tmp=New-TemporaryFile
Get-Content -Path $tmp.FullName | Should Be $null
$teststring ="any test string"
$teststring | Out-File -Path $tmp.FullName
Get-Content -Path $tmp.FullName | Should Be $teststring
Remove-Item $tmp.FullName -Force
```

lzybkr commented:
We can merge as soon as the conflicts are resolved.

kittholland commented:
Conflicts have been resolved, thanks.

lzybkr commented:
@kittholland - can you rebase? The PR now has a bunch of commits it shouldn't.

lzybkr commented on src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -2,7 +2,7 @@
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="� Microsoft Corporation. All rights reserved."
```
Revert this change please.

lzybkr commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -405,7 +374,7 @@ function Format-Hex
                     #           Path:
                     #           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
-                    #00000000   AA AB AC AD AE AF B0 B1 B2 B3 B4                 �����������
+                    #00000000   AA AB AC AD AE AF B0 B1 B2 B3 B4                 �����������
```
Revert this change please.

lzybkr commented on src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1:
```diff
@@ -418,9 +387,9 @@ function Format-Hex
-                    # 00000000   AA 00 00 00 AB 00 00 00 AC 00 00 00 AD 00 00 00  �...�...�...�...
-                    # 00000010   AE 00 00 00 AF 00 00 00 B0 00 00 00 B1 00 00 00  �...�...�...�...
-                    # 00000020   B2 00 00 00 B3 00 00 00 B4 00 00 00              �...�...�...
+                    # 00000000   AA 00 00 00 AB 00 00 00 AC 00 00 00 AD 00 00 00  �...�...�...�...
```
Revert this change please.

lzybkr reviewed (commented):
Some unicode characters got messed up, please fix.

kittholland committed:
Rewrite New-TemporaryFile in C#

Literal translation of New-TemporaryFile function to a C# cmdlet,
other than moving logic from Begin to End.

iSazonov commented:
@kittholland Thanks for your contribution!
